### ISSUE TITLE
Need to know the version tag for a StudyProtocolSnapshot

### ISSUE THREAD

**Issue by bardram** [cph-cachet/carp.core-kotlin#446]
In order to show it to the user in the web portal, we also need the version tag - which currently is not part of the snapshot. 

See >> https://github.com/cph-cachet/carp.core-kotlin/blob/develop/carp.protocols.core/src/commonMain/kotlin/dk/cachet/carp/protocols/application/StudyProtocolSnapshot.kt

**Comment by Whathecode** [comment]
I would refrain from adding it within `StudyProtocolSnapshot`, because protocol snapshots right now carry an "identity" which shouldn't change if the version tag changes.

But, I understand the need. Maybe `ProtocolService.getAllForOwner` should return a `VersionedStudyProtocol`, which contains `StudyProtocolSnapshot` and `ProtocolVersion`.

**Comment by Whathecode** [comment]
> But, I understand the need. Maybe ProtocolService.getAllForOwner should return a VersionedStudyProtocol, which contains StudyProtocolSnapshot and ProtocolVersion.

@bardram Would this fulfill the current use case?

So a new type, only returned by the protocol subsystem, would be introduced.

I presume the use case is along the lines of: as a researcher, I want to see which version the protocols I manage are in, so that I can make sure the right version is used when creating new studies with it.

Again, you can see that already through a separate call. And, the `getAllForOwner` endpoint returns the _latest_ version, which is what I expect you are generally interested in, if not, you can use the version history call. But, it does make sense to display to the user what this latest version is so they feel confident it matches expectations.

**Comment by yuanchen233** [comment]
I believe `GetAllForOwner`, 'GetVersionHistoryFor', and `GetBy` together serves all the functionalities needed in ProtocolService.

 > But, I understand the need. Maybe ProtocolService.getAllForOwner should return a VersionedStudyProtocol, which contains StudyProtocolSnapshot and ProtocolVersion.
> 
This would only be useful when a researcher 'owns' a lot of protocol, which is a rare case I would say.

After some discussion, the current use case would falls under StudyService. The `version` field desired is in `StudySnapshot.protocolSnapshot`: As a researcher, I want to know which version of the protocol is being used in the current study, without look into the content of the protocol snapshot.

This would be useful for situations where a protocol is being modified multiple times, with minor revisions, and studies are made with different versions of the same protocol. Or different versioned protocol is used in different control groups, managed in different studies.

Also, if protocols and studies are managed separately, it will be hard to see details of the protocol used by a study. For example, carp-researcher-portal visualise the content of a protocol in `Protocol Section`, it can get different version of the protocol using `getBy`. But a study does not keep track of which version of the protocol it uses, only 'latest version when creating the study', so we have to visualise the protocol used by the study in 'Study Section'.

In general, the 'version' of a protocol is treated like a new protocol, but the researcher is responsible for managing them. It makes sense of replacing 'myPtorocolV1.0, V1.1 ...' into one protocol with version tags, but then they should also be differentiated by name and tag, where the `version tag` part is missing.

**Comment by Whathecode** [comment]
> I believe GetAllForOwner, 'GetVersionHistoryFor', and GetBy together serves all the functionalities needed in ProtocolService.

True. It would simply be a shortcut to remove the need to do one extra endpoint call. It does feel a bit like premature optimization over keeping endpoints focused/simple.

> After some discussion, the current use case would falls under StudyService. The version field desired is in StudySnapshot.protocolSnapshot: As a researcher, I want to know which version of the protocol is being used in the current study, without look into the content of the protocol snapshot.

Thanks! That's a nice user story explaining the intended need.

I still don't think this means it belongs in `StudyProtocolSnapshot`. As you say, it is related to added requirements managed by the _studies_ subsystem. So we need to find a way to fit it in there. That's fairly trivial in the current state, in which a protocol is "locked in" once a study goes live. We can simply store it in the study at that point in time. Only for more advanced future scenarios (trial runs within a study, after which the protocol is updated, etc.) would we need to do something more involved. That's also _exactly_ why it belongs in the study subsystem and not in a type owned by the protocols subsystem. The main reason for change lies with how you want to/intend to support protocols _in studies_.

So, why not in `StudySnapshot`?

**Comment by yuanchen233** [comment]
Yes, it could simply be a `protocolVersion: ProtocolVersion` field for now.

> Only for more advanced future scenarios (trial runs within a study, after which the protocol is updated, etc.)

Interesting! and also complex, everything depends on the protocol, this will be a soft reset/re-deploy for the study.

We also keep a copy of `studyProtocolSnapshot` in other places tho: `Recruitment`, `StudyDeployment`.. we should also consider adding this `protocolVersion` to them, if we decided to go for this approach.

If so, we could have a study running, and deploy deplyments with different protocol version.

**Comment by Whathecode** [comment]
> We also keep a copy of studyProtocolSnapshot in other places tho: Recruitment, StudyDeployment.. we should also consider adding this protocolVersion to them, if we decided to go for this approach.

Since `Recruitment` reflects the "participants" side of things in a study, I doubt it. But, we can discuss if/when the need arises.

`StudyDeployment` unlikely, and even more likely to be avoided. This is the deployment subsystem, which we should really keep devoid of study management things which instead should be layered on top. The deployment subsystem knows nothing about "studies". If a study subsystem implementation ever decides to run different protocol versions within the same study, the linking of which deployment has which version should happen there.

It feels naturally at home in `StudyService`. And, checking in a bit more detail, probably `StudyDetails` would be the place to expose it (through the existing `getStudyDetails()`), and `StudySnapshot` to persist it.

**Comment by Whathecode** [comment]
After a meeting with @yuanchen233, we concluded that the easiest thing for now is to simply store the protocol version in the platform-specific implementation, i.e., CAWS.

CARP core, as is, doesn't have a baked in assumption that protocols in the protocols subsystem are immutable (not considering whether that's a feature we do need for now). I.e., even if you link back to a protocol version in the protocols subsystem, it may have been updated since the study started (e.g., see the mutable `updateParticipantDataConfiguration` call). Making _every_ edit in the protocols subsystem immutable would place an inordinate amount of overhead on the protocols subsystem, without a very clear use case for it. It is the _studies_ subsystem that "locks in" a study and which decides to no longer mutate a given study protocol. There is no need to store a new version for every single edit before that. And, if you want a clear overview of all "versions" of the protocol, having a version per edit also obscures the real intent/requirement. The "versions" use case is tailored to "pilot", "round 1", "round 2", etc.

Furthermore, neither is it a requirement in the current CARP core studies subsystem that a protocol of a study _exists_ in the protocols subsystem. It should be possible for protocols to come from elsewhere. Protocols are copied over in full, and their identity is their whole definition. The protocol ID is only there for semantic traceability, e.g., to find other versions.

We _could_ conceive a `ProtocolsService.lookupProtocol( StudyProtocolSnapshot )` endpoint, which goes through all protocols with the given ID registered in the protocols subsystem and finds an exact match. Similar functionality can already be implemented by the client by calling `getVersionHistoryFor` and doing the comparison client side. The main reason for a `lookupProtocol` call would be a performance optimization to prevent having to return _all_ protocol versions to the client, and to be able to optimize this on the backend through a hash table lookup. But, this likely is premature at this time. And, it could _still_, in theory, return multiple protocols, given that nothing in the protocols subsystem restricts adding the same snapshot under different versions.
